[WIP][#2810] Resource Landing Page: Reorganize Elements For Clarity and Easier Access

Maurier opened the pull request:
This PR contains updates to the resource landing page described in #2810, #2924, #2813, #2763, #2731. This work is currently deployed at https://riverdev.hydroshare.org where it can be tested.

- Removed tabular interface and reorganized the content.
- Added profile preview for resource landing page Owners listing
- Added preview for resource landing page Authors listing
	- Editing now is done in a modal window
	- Reordering is now done in placeby dragging and dropping
	- Added dragging placeholder indicator
- Added Highlights section to the resource header
	- Ability to see +1s and +1 the resource
	- Ability to see resource views count
	- Ability to see number of downloads for the resource
	- Ability to see number of comments and provide link to comments section
	- Exposed Sharing status to this section
- Added side panel in Content section to see file type metadata
	- Added toggle to show/hide side panel
- Adjustments to Google Maps UI in Spatial Coverage section.
	- Recenter button has been adjusted in size and is now properly rendered as a map component, rather than drawing it on top of the map
- Added README section
- Reorganized and restyled Set Content Type menu items
- Removed page reloads when creating File Aggregations
	- Resource level metadata is updated automatically
- Removed page reloads when removing File Aggregations
	- Resource level metadata is updated automatically
- Removed page reloads when uploading files
	- Resource level metadata is updated automatically
- Removed page reloads when using the checkbox to change Download Requirements
- Added Resource Size to resource landing page header
- Assign minimum with to section list view so it no longer cuts off file names.
- Added Content Types to resource landing page header
- Fix for unnecessary horizontal scroll bar appearing in Create Resource page
- Various bug fixes concerning Download Zipped functionality
- Enabled coordinates picker in file metadata
- Moved Download All Content button to File Browser's toolbar
- Moved iRODS upload functionality to File Browser's toolbar
- Moved Quota Holder content to Manage Access panel
- Restyled content inside Manage Access panel
- Fixed a bug in Manage Access panel where searching for a user or group to invite would display the autocomplete outside of the input area.
- Added logic to toggle Publish Resource button when access is changed
- Added default message to File Browser when displaying an empty directory
- Publish Resource button now shows up for Owner users, disabled if resource is not public
- Code cleanup and optimization
- Various color and font changes
- Added text to Save Changes button to indicate when a request is being processed
- Changed the text on the options that appear when adding and displaying Relations
- Added logic to render modal items only when they are needed.
	- This helps prevent web crawlers from unnecessarily triggering forms
- Improved logic in charge of enabling and disabling File Browser's menu items based on selected files

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [ ] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]

sblack-usu commented:
@Maurier - you have 6 failing tests and 11 lines that need to be fixed up for pep8 compliance.

Maurier commented:
@sblack-usu Noted, I'll be updating those soon.

sblack-usu commented on django_irods/views.py:
```diff
@@ -354,6 +354,9 @@ def download(request, path, rest_call=False, use_async=True, use_reverse_proxy=T
 
         if not res.is_federated:
+            # track download count
+            res.update_download_count()
```
agh, I loathe this download method.  We keep adding complexity and dependencies to it and we already know it's a mess.  Perhaps this tracking could be moved to the front door of when a download link is hit instead of in here when the file is served.  I'd like to minimize how much django_irods uses hydroshare objects.

pkdash replied:
The concern that I have moving the download tracking inside the initial view function in hs_core is that it would count download even if the download fails with the django_irods. I guess it may not be that critical if the download count is not very accurate. I am OK with moving it out of django_irods.

sblack-usu replied:
@horsburgh - any thoughts on how accurate you would like the download count to be?  @pkdash is correct that by moving this count to the front door, it would still count failed downloads.

horsburgh replied:
@sblack-usu - I guess it should be as accurate as we can make it. A failed download attempt isn't really a download. Do we anticipate a lot of failed download attempts? If this artificially inflates the download count a little, it probably isn't a big deal because a failed download is still someone wanting to download. But, if we anticipate this happening a lot, we should probably try to prevent it.

pkdash replied:
@horsburgh I don't anticipate lot of download failures in general. However, if a resource or file type gets corrupted for some reason, then any attempt to download resource bag, files or folder for that resource would most likely fail until the resource or file type gets fixed. 

@sblack-usu It seems the view function in hs_core that supports file download is not used for resource bag download. For resource bag download the download view function in django_irods is directly used. I guess for now I have to keep the download count in django_irods.

sblack-usu replied:
ok

sblack-usu commented on hs_core/models.py:
```diff
@@ -2252,16 +2277,71 @@ def has_required_content_files(self):
+        """Returns True if the resource has a readme.md or readme.txt file at the root, otherwise
+        False."""
+
+        return self.readme_file is not None
```
can you remove the is not None?

sblack-usu commented on hs_core/models.py:
```diff
@@ -2252,16 +2277,71 @@ def has_required_content_files(self):
+                return {'content': markdown(self.readme_file.read().decode('utf-8')),
+                        'file_name': self.readme_file.file_name}
+            else:
+                return {'content': self.readme_file.read(), 'file_name': self.readme_file.file_name}
```
Why do we need to support txt files as well?

pkdash replied:
@horsburgh probably knows the answer.

horsburgh replied:
@sblack-usu and @pkdash - when we originally discussed readme functionality, we started with readme.txt, but then people wanted to do rich text formatting, so we added the ability to do markdown via readme.md. I think it's still valid for someone to have a readme.txt file that does not contain markdown formatting. So, the rule is, a readme.md or a readme.txt file can be uploaded, and either one will be displayed on the page if only one file is uploaded. If neither are uploaded, nothing is displayed. If both are uploaded, the readme.md gets preference over a readme.txt file for display.

sblack-usu replied:
My only argument for removing txt file support is to keep things simpler both in the logic and usability.  It's not much more logic to support txt files, so not a huge deal, but I'm not sure we really gain much by supporting both and then we also have to clarify to users when one file is displayed over the other.  I'm not going to hold anything up over this, just voicing my opinion that I think it's sufficient to only support markdown.

horsburgh replied:
@sblack-usu - For simplicity sake, I would be OK with letting the rendering of readme.txt go as long as we have buy-in from the team. There's sort of a history associated with readme.txt as a place to put documentation of scientific datasets. I'm not sure a lot of users will know what a ".md" file is (although it is becoming more common as more people are using GitHub). Users could still upload a readme.txt file, it just wouldn't be rendered on the page. If we don't support both, we likely need some indication somewhere that says - "Upload a readme.md file with documentation for your resource. The text will be displayed on the resource landing page and formatting is supported via Markdown." 

Or something like that...

sblack-usu commented on hs_core/models.py:
```diff
@@ -38,7 +39,7 @@
 from mezzanine.pages.managers import PageManager
 
-from dominate.tags import div, legend, table, tbody, tr, th, td, h4
+from dominate.tags import div, legend, table, tbody, tr, th, td, h4, br
```
when and why did we start using dominate?  I know it's been around for awhile, but i'm curious why we started using it.

pkdash replied:
I find it useful for dynamically generating html snippets from python. Otherwise, I guess we would have to generate these html snippets in JavaScript. I preferred to do it in python as it was easier for me to do it in python vs JS.

sblack-usu replied:
I guess I don't see why we couldn't just use django templates to create the snippets.  For example, couldn't the get_html method on line 1355 that was modified below just have two templates setup with that html structure, then you could just return the appropriate template based on the logic on type and pass in the context?

I definitely don't want pure JS.  @cuahsimarko I looked into jinja2 a little bit and it seems to be just a newer version of Django template... It's not clear to me what it would add.

I like the looks of dominate as a developer that doesn't spend much time in the frontend... however I think it's less appealing to frontend devs like @Maurier and @engineSound and I'm worried it might cause some bleeding between the front and back ends.

pkdash replied:
@Maurier and @engineSound As you can see @sblack-usu has raised a valid concern that front end developers may not feel comfortable the way I am generating html code using python. All file type/aggregation level metadata related html is generated in python. If you prefer that we should be using Django template to write the html code for filetype metadata then I would suggest we create a new issue for this conversion and one of you can do this conversion. I would be happy to help you in that conversion.

sblack-usu commented on hs_core/models.py:
```diff
@@ -1668,6 +1666,33 @@ class MyResourceContentType(pages.Page, hs_core.AbstractResource):
+    def download_count(self):
+        return int(self.extra_data.get('download_count', '0'))
+
+    def update_view_count(self, request):
```
I don't get this method. I see three routes here:
1. the request has resource_ids
  a. the resource id is not in the resource_ids - updates the view count
  b. the resource id is in the resource_ids - the view count is not update (why not?)
2. the request does not have resource_ids - the view count is updated.

I'm not following why this method does and does not update the view_count.  And I believe the request object has no business being in here.

sblack-usu commented on hs_core/models.py:
```diff
@@ -1668,6 +1666,33 @@ class MyResourceContentType(pages.Page, hs_core.AbstractResource):
+        cur_count = self.extra_data.get(key, '0')
+        new_count = int(cur_count) + 1
+        self.extra_data[key] = str(new_count)
+        self.save()
```
Why did you decide to use extra_data for these counters?  Looking at extra_data, we're not actually using it anywhere up until now and I'm not sure this is the best use for it.  The view count is arguably the most updated field on a resource.  I think it deserves it's own field for quicker access/update.  I get that it's not a huge difference in overall time, but I'd expect it takes a whole lot more time to access the extra_data field, look up the key, convert it to integer, increment the integer, convert it back to string, insert into the map and save the field, than it would be to just increment an integer field.

pkdash replied:
I agree that using model level individual fields for the counters would be simpler. Initially I was planning to do that. Then I thought there may be need for additional such counters in the future - leading to model level modifications and migrations. The 'extra_data' field has been used by collection resource type. I will make these counters model level fields.

sblack-usu commented on hs_core/models.py:
```diff
@@ -2252,16 +2277,71 @@ def has_required_content_files(self):
+        'readme.txt' or 'readme.md' (filename is case insensitive). If no such file then None
+        is returned. If both files exist then resource file for readme.md is returned"""
+
+        res_files_at_root = [f for f in self.files.all() if f.file_folder is None]
```
We don't want to look through all the files for a readme each time the page loads.  These types of things add up and contribute to the long page loading times.

sblack-usu commented on hs_core/models.py:
```diff
@@ -2252,16 +2277,71 @@ def has_required_content_files(self):
+        else:
+            if self.readme_file.extension.lower() == '.md':
+                return {'content': markdown(self.readme_file.read().decode('utf-8')),
+                        'file_name': self.readme_file.file_name}
```
self.readme_file is called 4 times in this method.  That's 4 times the readme file is identified by looking through all the files in the resource.

sblack-usu replied:
I see you updated this logic to put the readme_file into a variable.  To take this further, I think the only time the readme_file would change is when a readme file is uploaded or deleted.  Do you have thoughts on adding a readme_file pointer to the model and just using that instead of looking it up every time?

sblack-usu commented on hs_core/models.py:
```diff
@@ -2252,16 +2277,71 @@ def has_required_content_files(self):
         return logical_files_list
 
+    @property
+    def aggregation_types(self):
```
I don't think this is used anywhere.

pkdash replied:
It is used in a template 'resource-header.html'

sblack-usu replied:
ah, I guess Pycharm doesn't look in templates for usage.

sblack-usu commented on hs_core/page_processors.py:
```diff
@@ -117,10 +117,15 @@ def get_page_context(page, user, resource_edit=False, extended_metadata_layout=N
 
     # user requested the resource in READONLY mode
     if not resource_edit:
+        content_model.update_view_count(request)
```
maybe call it increment_view_count and don't pass in request?

sblack-usu commented on hs_core/views/utils.py:
```diff
@@ -1265,3 +1266,29 @@ def get_coverage_data_dict(source, coverage_type='spatial'):
         return temporal_coverage_dict
+
+
+def get_resource_metadata(resource):
```
Is this so we don't have to do a full page refresh and adding files and setting a file type?

sblack-usu commented on hs_file_types/models/generic.py:
```diff
@@ -57,10 +57,10 @@ def get_html(self):
-            with root_div:
-                h4("No file level metadata exists for the selected file.")
-            html_string = root_div.render()
+            html_string = '<div id="#fb-metadata-default" class="text-center text-muted" ' \
```
why did you stray away from dominate here and use a string?

sblack-usu commented on hs_file_types/models/geofeature.py:
```diff
@@ -85,11 +85,11 @@ def get_html_forms(self, datatset_name_form=True):
-            with div(cls="col-lg-6 col-xs-12 col-md-pull-6", style="margin-top:40px;"):
+            with div(cls="content-block"):
                 div("{% crispy spatial_coverage_form %}")
-            with div(cls="col-lg-6 col-xs-12"):
```
I like that you're removing styling from the html building.  @engineSound - Could you review the templates, dominate and changes to organizing styling?  Also, what are your thoughts on dominate vs templates?

sblack-usu commented on hs_file_types/models/raster.py:
```diff
@@ -91,26 +92,22 @@ def get_html_forms(self, dataset_name_form=True, temporal_coverage=True, **kwarg
-                                               cls="btn btn-primary pull-right btn-form-submit",
-                                               style="display: none;")
-                        div("{% endfor %}")
+                div("{% crispy orig_coverage_form %}", cls="content-block")
```
ok, I realize you just did some formatting changes, but this is the first time I've noticed we're using template language inside dominate... It makes me uncomfortable seeing this, but I'm going to have to chew on it a bit to better understand the implications.  @engineSound @cuahsimarko - Do you have any concerns or praises for doing this?

cuahsimarko replied:
I haven't worked with dominate before. I've seen jinja2 approach as well as a fully separated backend/API - ReactJS frontend. I asked in a previous comment what our other options might be here.

sblack-usu replied:
@engineSound - would you please chime in?

engineSound replied:
The segment of code has many levels. It is challenging to read and maintain. Regardless of what technology is being used here, I would recommend re-engineer this section.

sblack-usu commented on hs_file_types/models/reftimeseries.py:
```diff
@@ -495,26 +495,26 @@ def get_html(self):
-            with root_div:
-                h4("No file level metadata exists for the selected file.")
-            html_string = root_div.render()
+            html_string = '<div id="#fb-metadata-default" class="text-center text-muted" ' \
```
Again, why move away from dominate to a string?

pkdash replied:
In fact this error message is not needed as ref time series will always have metadata. Removing the string and check for metadata.

sblack-usu commented on hs_file_types/models/timeseries.py:
```diff
@@ -106,40 +106,45 @@ def get_html(self, **kwargs):
                     if site:
                         legend("Site")
                         site.get_html()
+                        br()
```
a line break is styling... I don't think it belongs in here.

sblack-usu commented on hs_file_types/views.py:
```diff
@@ -82,8 +83,8 @@ def set_file_type(request, resource_id, hs_file_type, file_id=None, **kwargs):
         response_data['message'] = msg
-        spatial_coverage_dict = get_coverage_data_dict(res)
-        response_data['spatial_coverage'] = spatial_coverage_dict
+        resource_metadata = get_resource_metadata(res)
```
I don't know if this belongs here.  I feel like a better pattern would be to send this metadata over in it's own ajax call.  That way, when we add more functions that require the metadata to be refreshed, the frontend can make the call instead of updating the server to include this data with each call?

cuahsimarko commented:
Test is failing FAIL: test_set_resource_flag_make_not_shareable

sblack-usu commented on irods_browser_app/static/js/irods_signin.js:
```diff
@@ -56,16 +70,28 @@ function irods_login() {
-                $("#btn-select-irods-file").toggleClass("hidden", false);
-                $("#irods-sel-file").text("No file selected.");
+                if ($("#irods-group").length) {
+                    // Logic for resource landing page
```
the logic here and in create resource page are duplicated at least three times in this file... It's probably time to put them into a method and call it.

Maurier replied:
I agree. But this could be an improvement we should handle in a subsequent ticket.

sblack-usu replied:
I'm not sure it makes sense to kick this can down the road when it has all been touched in this branch.  I think the time to address it is now unless you have a ticket that has this work in it's scope.

cuahsimarko replied:
I vote to get this code out the door.

Maurier replied:
@sblack-usu This will be implemented in a separate ticket as it requires more than consolidating the logic in .js files. Both uses differ in their template and so I think a consolidation of this functionality is the best way to go. Created a further ticket to address it: https://github.com/hydroshare/hydroshare/issues/3095

sblack-usu commented on hs_geo_raster_resource/models.py:
```diff
@@ -160,27 +160,25 @@ def get_html_form(cls, resource, element=None, allow_edit=True, file_type=False)
-                        td(self.value.get('projection_string', ''))
+            div('Coordinate Reference System', cls='text-muted')
+            div(self.value.get('projection', ''))
+            br()
```
found more line breaks

Maurier replied:
Noted.

sblack-usu commented on theme/static/css/hs-file-icons.css:
```diff
@@ -98,7 +98,9 @@
 
 .fb-view-list .fb-stack .fa-file-o {
-	font-size: 18px;
+    font-size: 18px;
```
@engineSound - could you review the frontend html structures and styling?

engineSound replied:
@sblack-usu The html and css overall looks good to me. Just some minor comments here and there. Maurier has a good work on the html and css. On the same line, we need to discuss separately where and how we encapsulate html snippets to avoid them to spread everywhere #3097.

sblack-usu commented on theme/static/css/hydroshare_core.css:
```diff
@@ -1369,6 +1497,25 @@ body > .main-container {
+}
+
+.custom-btn-toolbar {
+    display: flex;
```
spaces here and tabs above?  I'm seeing this all over this css file... it's probably elsewhere too.  I think we need to run these files through a lint.

Maurier replied:
I agree. We should probably make a separate ticket for this and enforce lint after that.

sblack-usu replied:
That's fine, please create the ticket so we can get it prioritized.

sblack-usu commented on theme/static/js/custom.js:
```diff
@@ -191,7 +176,185 @@ $(document).ready(function () {
+    $(".profile-preview").click(function () {
+        // Move the profile card below the clicked item
+        var profileCard = $(this).parent().find(".profile-card");
+        profileCard.css("top", ($(this).position().top + 30) + "px");
```
I don't do tons in the frontend, but shouldn't this type of thing be done by switching out classes?  I guess that might not be possible with the references to position and width?  I don't know, seems pretty messy.

Maurier replied:
Not in this case because we need to acccess the top() method which is a dynamic value.

sblack-usu commented on theme/static/js/hs_file_browser.js:
```diff
@@ -78,332 +101,376 @@ function formatBytes(bytes) {
+
+    // Initialize the action states
+    $.each(uiActions, function (i, val) {
+        uiActionStates[val] = $.extend(true, {}, initActionState);  // Deep copy
```
uiActionStates is an improvement from the flag variables that existed before 👍

Maurier replied:
I agree. A dictionary makes them easy to find.

sblack-usu commented on theme/static/js/hs_file_browser.js:
```diff
@@ -1243,7 +1300,7 @@ $(document).ready(function () {
 
                 this.on('sending', function (file, xhr, formData) {
-                    formData.append('file_folder', $("#upload-folder-path").text());
+                    formData.append('file_folder', "data/" + $("#upload-folder-path").text());
```
What/Where/Why did "data/" pop up in here and above in 1247?  I like to gripe about how we handle /data/contents/ path differently in the system.

Maurier replied:
When sending a request to the API we need to send the full path as it is implemented originally. But since the data folder is not accessible in the file browser we don't need to show it in the UI.

@pkdash I suggest we make the change accordingly to ommit the "data" part from the path. At least for request coming from the UI.

sblack-usu replied:
You should be able to omit data/contents as it is mostly invisible to the user outside of a bag.

pkdash replied:
@Maurier The back-end view function for file upload should work if the UI doesn't include '/data/contents' as part of the path. I assume you also want the call to refresh file browser for a given path not to include '/data/contents' as part of the path. That needs change in the corresponding view function at the back-end.

Maurier replied:
@pkdash, that's correct. Let's remove the 'data/contents' part from the path for the refresh method.

sblack-usu commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -110,17 +169,33 @@ function license_agreement_ajax_submit(event) {
             // page refresh is needed to update if license agreement popup to show or not prior
             // to download of files/bag
-            window.location = window.location.href;
+            // window.location = window.location.href;
```
delete?

sblack-usu commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -818,9 +905,9 @@ function set_file_type_ajax_submit(url, folder_path) {
-            var json_response = JSON.parse(result);
-            var spatialCoverage = json_response.spatial_coverage;
-            updateResourceSpatialCoverage(spatialCoverage);
+            // var json_response = JSON.parse(result);
```
delete?

sblack-usu commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -1142,7 +1250,7 @@ function get_irods_folder_struct_ajax_submit(res_id, store_path) {
 
             $("#hs-file-browser").attr("data-current-path", store_path);
-            $("#upload-folder-path").text(store_path);
+            $("#upload-folder-path").text(store_path.replace("data/", "")); // We don't show the data folder in the UI path
```
ah, here we go.  I say we don't show content as well and then update the server to no longer expect "data/contents" for file paths.

sblack-usu replied:
I've looked at the backend and it's enforcing the path to start with data/contents...  Should be easy enough to remove that and just prefix the path with data/contents if it's not passed in.

sblack-usu commented on theme/templates/accounts/account_login.html:
```diff
@@ -59,7 +59,7 @@ <h2 class="page-title">Sign In</h2>
             </form>
 
-            <p>{% blocktrans with request.GET.next as next %}If you don't have an account you can <a href="/sign-up/?next={{ next }}">join HydroShare</a> now.{% endblocktrans %}</p>
+            <p>{% blocktrans with request.GET.next as next %}<br>If you don't have an account you can <a href="/sign-up/?next={{ next }}">join HydroShare</a> now.{% endblocktrans %}</p>
```
should this line break be done in css instead?

sblack-usu commented on theme/templates/accounts/sign_up_form.html:
```diff
@@ -48,12 +48,10 @@
-
-  </div>
+        <div id="missing-data-div"></div>
+        <br>
```
again, don't think this belongs here.

sblack-usu commented on theme/templates/generic/includes/rating.html:
```diff
@@ -1,11 +1,30 @@
+            {% if rating_sum == 1 %}
+                <span>You +1 this</span>
+            {% elif rating_sum == 2 %}
+                <span>You and 1 other +1 this</span>
```
Doesn't the dynamic rating_sum doesn't cover this case?

sblack-usu replied:
I'm confused from my comment too!  I was just pointing out that I think you can remove this block as the else block will create the same message when rating_sum equals 2.

sblack-usu replied:
... Now I see what case this is covering, others is plural in the else block below.  Nevermind, nothing to see here.

sblack-usu commented on theme/templates/includes/author-link.html:
```diff
@@ -0,0 +1,18 @@
+<a data-toggle="modal" class="author-preview"
```
Isn't this an example of an html snippet using django templates?  (An alternative to using dominate or pure JS)

sblack-usu replied:
So you're in favor or django snippets instead of the alternate solutions we have scattered through the system?

Maurier replied:
Indeed. This offers great re-usability through Django template language, while providing all the flexibility of writing HTML directly.

pkdash replied:
In my opinion you can do more with Dominate than what you can do with django template. For example, instead of template filters you can use python to do all those magic. You have full power of python for generating html code. The html generated using dominate tags goes through django template engine before it is sent to UI. You can do whatever re-usability you can get in template language using dominate.  The only disadvantage I see using dominate is that front-end developers not familiar with python may find it difficult to maintain. For that reason only, I would agree that we should replace dominate code with django template code.

sblack-usu commented on theme/templates/includes/author-preview.html:
```diff
@@ -0,0 +1,85 @@
+                <h4 class="modal-title">Author</h4>
+            </div>
+            <div class="modal-body">
+                <table class="table" style="border: 1px solid #DDD;">
```
put styling in css?

sblack-usu commented on theme/templates/resource-landing-page/modals.html:
```diff
@@ -1,847 +1,412 @@
+                {#  Organization  #}
+                <div data-hs-user-type="organization" style="display:none">
+                    {# This code block should be uncommented after the auto complete for Existing HS Groups is working #}
+                    <!--<div class="modal-body">
```
remove commented code.

sblack-usu commented:
Talked to @dtarb about the readme files, and he would like to support txt files.  I looked to github to see what they're doing.  And they are doing the rules that @horsburgh layed out (supports both md and txt, gives display priority to md).

engineSound commented on hs_core/templates/pages/create-resource.html:
```diff
@@ -196,68 +196,66 @@ <h3>Drop files here or click to upload</h3>
+                                                class="glyphicon glyphicon-log-in"></span>&nbsp;&nbsp;Log in to iRODS...</a>
+                                    </div>
+                                    <br>
+                                    <input type="hidden" id="irods-username" name="irods-username" value="">
```
I like your articulate the id for each tag. This is a tremendous help for automation testing.

engineSound commented on hs_core/templates/pages/my-resources.html:
```diff
@@ -353,92 +353,95 @@ <h4 class="modal-title" id="myModalLabel">Create New Label</h4>
+                    </table>
+                </div>
+                <div class="modal-footer">
+                    <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
```
id field. Chrome will generate warning for missing ID.

Maurier replied:
@engineSound this button does not need an id attribute at the moment. If it is needed for automated testing I think the id should be added once the test is written.

engineSound replied:
Automation testing shouldn't touch source code.

engineSound commented on hs_core/templates/pages/my-resources.html:
```diff
@@ -353,92 +353,95 @@ <h4 class="modal-title" id="myModalLabel">Create New Label</h4>
+                    <div>Select resources to delete</div>
+                </div>
+                <div class="modal-footer">
+                    <button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button>
```
ID attribute. We should have id attribute for every html tag.

Maurier replied:
I disagree. I think id attributes should only be added if there is a need to specifically reference the element somewhere else.

sblack-usu replied:
I agree with @Maurier - only write code that we use.  If functionality that comes along that needs the ids, this can be added as needed.

engineSound replied:
selenium automation need the id for referencing UI element to emulate user interaction. I observed Neal doing this. Without ID, the automation code is hard to write and difficult to maintain. Maybe there is another way that help to automatically generate random id for each element to facilitate automation testing. I would discuss this with @ndebuhr separately.

engineSound commented on hs_model_program/templates/pages/modelprogramresource.html:
```diff
@@ -6,105 +6,101 @@
+                                {% else %}
+                                    <th class="text-muted">Computational Engine</th>
+                                {% endif %}
+                                <td><a id="permalink"
```
Suggest a different id.

Maurier replied:
@engineSound There are a bunch of tags with this id. I suggest we get rid of these attributes instead since they are not being referenced anywhere in the project that I am aware.

engineSound commented on hs_tools_resource/templates/pages/toolresource.html:
```diff
@@ -7,180 +7,159 @@
+
+                <h4>App-launching Aggregation Level URL Pattern:
+                    <a href='/terms#AppURLPattern' target='_blank'>
+                        <font size='3'>Help</font></a>
```
outsource this to CSS

pkdash replied:
Agree. For now I am changing this to a view function so that the UI can make a separate call to get this limited set of resource level metadata needed for updating the UI. Later (not in this PR) I think it would be better to change this view function to return data only for a specified metadata element - so that the UI can access data for any supported metadata element.

engineSound commented on theme/static/js/custom.js:
```diff
@@ -191,7 +176,185 @@ $(document).ready(function () {
+    $(".profile-preview").click(function () {
+        // Move the profile card below the clicked item
+        var profileCard = $(this).parent().find(".profile-card");
+        profileCard.css("top", ($(this).position().top + 30) + "px");
```
How about the view on mobile devices?

Maurier replied:
I tested this on mobile devices and it worked well too. I will give it another round of testing just to be sure.

engineSound commented on theme/templates/resource-landing-page/coverage.html:
```diff
@@ -1,50 +1,36 @@
+                    {% if spatial_coverage.units %}
+                        <div>
+                            <i class="fa fa-tachometer" aria-hidden="true"></i>
+                            <small>&nbsp;Coordinate Units:&nbsp;</small>
```
We can use CSS to replace the non-breaking space. Non-content related stuff should be considered to move to CSS.

Maurier replied:
Agree. Will change.

engineSound commented on theme/templates/resource-landing-page/coverage.html:
```diff
@@ -252,55 +241,51 @@ <h4>Spatial:</h4>
+                                        {% if not cm.resource_type == "CompositeResource" or not cm.has_logical_spatial_coverage %}
+                                        style="display: none"
+                                        {% endif %}>
+                                    <i class="fa fa-calculator" aria-hidden="true"></i> Set spatial coverage from content files
```
If we have time, we can move the white space character into CSS.

Maurier replied:
Agree. Will change too.

engineSound commented:
Resource Landing Page redesign has fixed issues that were previously filed. To prevent the issues from loss of tracking, here they are. I will come back to update this. 
# 2945

hydrobuild commented:
Can one of the admins verify this patch?

pkdash commented on hs_composite_resource/tests/test_readme_file.py:
```diff
@@ -0,0 +1,232 @@
+            groups=[self.group]
+        )
+        # create readme files
+        self.readme_txt = "readme.txt"
```
I think trying to make code bit compact here may result into less readability.

pkdash commented on hs_core/views/__init__.py:
```diff
@@ -10,7 +10,6 @@
 from django.contrib.auth.models import Group, User
 from django.contrib.sites.models import Site
 from django.contrib import messages
-from django.contrib.messages import get_messages
```
Most of the changes related to 'resource sharing status' change are due to we no more using django 'messages' framework to propagate message related to status change to the UI. The same messages are now part of the json response object.

pkdash commented:
@sblack-usu Can you look into the one failing test? I assume you wrote this unit test. I think I know why it is failing. But not sure how to fix it. I think this line (https://github.com/hydroshare/hydroshare/blob/2810-reorganize/hs_core/tests/test_dx_migration.py#L49)  is problematic as we are not using the 'BaseResource' model based on the 'old_apps' when creating  the resource. Since there is now a new migration file as part of this PR causing this unit test to fail.

Maurier commented on theme/static/css/jquery.dataTables.min.css:
```diff
@@ -0,0 +1 @@
+table.dataTable{width:100%;margin:0 auto;clear:both;border-collapse:separate;border-spacing:0}table.dataTable thead th,table.dataTable tfoot th{font-weight:bold}table.dataTable thead th,table.dataTable thead td{padding:10px 18px;border-bottom:1px solid #111}table.dataTable thead th:active,table.dataTable thead td:active{outline:none}table.dataTable tfoot th,table.dataTable tfoot td{padding:10px 1 ...
```
We are using this script in multiple pages. It handles functionality such as table column sorting and filtering. This branch simply moves the script to our own files to serve it more reliably, since I remember at least one time where the previous hosting service was unavailable briefly in the past.

![image](https://user-images.githubusercontent.com/2448568/51124635-ffe53980-17db-11e9-8616-0e7b005235c1.png)

sblack-usu replied:
Those are all serving the script on the page, but I don't see where the dataTables jquery library is actually being used though.  datatables.net says you attach it to a table with 
```
$(document).ready( function () {
    $('#myTable').DataTable();
} );
```
A text search in the project show DataTable() only found in the jquery library itself.  There may be another way to call it that I'm not seeing... Could you point that out for me?

Maurier replied:
Here it is. This script has the functionality that is reused. 
https://github.com/hydroshare/hydroshare/blob/2810-reorganize/theme/static/js/hs_resource_table.js#L42

sblack-usu commented:
> 
> 
> @sblack-usu Can you look into the one failing test? I assume you wrote this unit test. I think I know why it is failing. But not sure how to fix it. I think this line (https://github.com/hydroshare/hydroshare/blob/2810-reorganize/hs_core/tests/test_dx_migration.py#L49) is problematic as we are not using the 'BaseResource' model based on the 'old_apps' when creating the resource. Since there is now a new migration file as part of this PR causing this unit test to fail.

It's safe to go ahead and delete that test.  The migration will never be run in production again, so there is no danger in future changes causing the migration to break.

sblack-usu commented on Dockerfile:
```diff
@@ -3,6 +3,8 @@ MAINTAINER Phuong Doan pdoan@cuahsi.org
 # inplaceedit in pip doesn't seem compatible with Django 1.11 yet...
 RUN pip install git+https://github.com/theromis/django-inplaceedit.git@e6fa12355defedf769a5f06edc8fc079a6e982ec
+RUN pip install \
+  markdown
```
This has got to go!  I just cleaned this file up... put it in hs_docker_base please.

Maurier replied:
@sblack-usu Could you point me to the line in hs_docker_base where I can add this package?

sblack-usu replied:
https://github.com/hydroshare/hs_docker_base/blob/82bf23bbc02f67d0a4e7c503ed69a2e5c338e472/requirements.txt

cuahsimarko commented:
@sblack-usu  @Maurier 
I'm responding to Scott's earlier feedback summary:
1. I think we should use Django templates and not involve yet another library. Code should be refactored before this PR is merged. (A separate comment about DRYing up some functions can wait, though)
4. Someone with PyCharm should go into the offending files and use its automated functionality in a subsequent PR to hit a bunch of the front-end and backend files for spacing/formatting. https://www.jetbrains.com/help/pycharm/reformatting-source-code.html
Related: I propose we require PyCharm and Visual Studio code if you are going to do backend/frontend dev. People, who use anything else should be held to strict scrutiny/standards around PEP8 and code formatting.

Maurier commented on theme/static/js/hs_file_browser.js:
```diff
@@ -1693,23 +1750,30 @@ $(document).ready(function () {
-            $("#license-agree-dialog-file").modal('hide');
+    // Download zipped method
+    $("[data-fb-action='downloadZipped']").click(function () {
+        if ($("#hs-file-browser").attr("data-agreement") === "true") {
```
@sblack-usu As far as I know we do have the agreement for zipped downloads, and I remember cleaning up the logic and fixing a bug that was causing it to not appear. Pinging @horsburgh and @dtarb for confirmation.

sblack-usu replied:
ok, then please just verify a download agreement appears.

Maurier replied:
Verified. I tested this today.

Maurier commented on hs_collection_resource/templates/pages/collectionresource.html:
```diff
@@ -338,7 +338,7 @@ <h4 class="modal-title">Deleted Resources</h4>
 {% block extra_js %}
     {{ block.super }}
-    <script type="text/javascript" src='//cdn.datatables.net/1.10.10/js/jquery.dataTables.js'></script>
+    <script type="text/javascript" charset="utf8" src="{{ STATIC_URL }}js/jquery.dataTables.js"></script>
```
@engineSound Right now we include this attribute in all our script tags. Unless we are ready to make the change in all files, we should avoid doing this now in the name of consistency. I propose we do this at the same time we take care of the lint styling in CSS files.

engineSound replied:
sure.

Maurier commented:
@sblack-usu, @engineSound I left some comments in this round of review.

sblack-usu commented:
> 
>     2. Someone with PyCharm should go into the offending files and use its automated functionality in a subsequent PR to hit a bunch of the front-end and backend files for spacing/formatting. https://www.jetbrains.com/help/pycharm/reformatting-source-code.html
>        Related: I propose we require PyCharm and Visual Studio code if you are going to do backend/frontend dev. People, who use anything else should be held to strict scrutiny/standards around PEP8 and code formatting.

We already are holding python files to strict pep8 formatting, it's the other file types that need to be cleaned up with some standards in place.  No reason to force others to use an IDE, we can enforce standards at the PR with the jenkins job like we do for pep8.

Maurier commented:
@horsburgh The mardown python package we are using is an implementation of John Gruber’s Markdown (https://daringfireball.net/projects/markdown/). Which should consider linking to this article instead of GitHub's in the help message above the file browser as it is more accurate.

horsburgh commented:
I guess I don't care so much which Markdown syntax reference we use. I've used the daring fireball one many times. My only consideration with that link is its longevity.  If you don't think that is a problem, add the URL link that you think is best.

Maurier commented:
@cuahsimarko 
While I agree it would be great to transcribe code to Django templates, I don't want to hold the functionality in this PR to address it. I think the sooner we get this PR merged, which is already quite large, the better. This way we avoid potential merge conflicts in the making.

I will create separate tickets for this and some other issues pointer out in this thread once this works is merged.

Maurier commented:
@sblack-usu Here is the feedback requested from the reviews. I think we have addressed all review comments at this point. If there is anything we might have missed please let us know.

> 1. html snippets are produced as a string in python, string in JS, dominate in python, and django templates... We need to decide what the standard is here and move forward with it.  I'm in favor of django templates, but I'm happy to be convinced otherwise if someone can make a case for it.


I think most of us agree on django templates for now. I will be opening a ticket for further discussion and transcribing tasks (https://github.com/hydroshare/hydroshare/issues/3093).

> 2. displaying/passing /data/contents path for the file browser.  I don't see why we can't omit /data/contents from the root of the path when requesting the file browser from the server.  Also, there were changes on the frontend that don't include data in the path anymore, causing data to be concatenated to the path before a request to the server is made.  Eliminating the /data/contents/ requirement would clean things up.


@pkdash and I have cleaned things up and "data/contents/" is no longer being passed in the client's requests or served in the responses.


> 3. There are some questionable decisions in here for organizing code.  I pointed out some styling in the html snippets and some styling in javascript.  We need to be careful to keep each layer separate and generally keep our code clean (delete commented out code).

I did quite a bit of cleanup on HTML and .js files. I will keep an eye on inline styling in HTML files and progressively move those into CSS.



> 4. A lot of the frontend files have mixed formatting (mixed use of tabs/spaces and inconsistent indentation).  We need to settle on a lint for html/css/js files.

I agree with this observation. I'll also create a ticket for further discussion on this as this. https://github.com/hydroshare/hydroshare/issues/3094


> 5. There appears to be an unused jquery data column library added to the project.  It needs to be removed if it's not being used.  If it is, please show me where it is being used and convince me we need it.

This has been clarified.


> 6. ~readme.md and readme.txt.  I'm in favor of getting rid of support for displaying readme.txt... seems like I sort of have @horsburgh convinced, we just need a solid decision to go forward with.~

We will keep supporting .txt files as agreed in the discussion above.


> 7. I have a few unanswered questions for some decisions in the backend, most of them minor. @pkdash has been responsive with many of the questions considering his limited time.

I believe @pkdash has addressed all reviews related to backend. 


> 8.  The jenkins build is failing.

At this point the tests have been updated and are no longer failing.

sblack-usu commented:
@Maurier - I'll approve once that pip install for markdown is removed from the Dockerfile.  This should be in hs_docker_base.  If you need more guidance on how that is setup, I'm happy to talk you through it.

Maurier commented:
@sblack-usu, Sounds good! @pkdash was taking care of it yesterday so hopefully we'll get that in soon.

sblack-usu commented:
@Maurier @pkdash - I'm building a new hs_docker_base on dockerhub right now.  Once it's done, you can update "release-1.9.1" to "release-1.9.2", remove the pip install markdown and commit.

Maurier commented:
@pkdash, yes. I'll push the change once @sblack-usu is ready.

Maurier commented:
@sblack-usu Let me know if we have the all clear to merge.

sblack-usu commented:
@Maurier - go ahead and do the honors of merging
